bundles/org.openhab.binding.luftdateninfo/src/main/java/org/openhab/binding/luftdateninfo/internal/handler/HTTPHandler.java

```diff
@@ -0,0 +1,182 @@
+                case 401:
+                case 404:
+                    logger.info("Sensor response: {}", httpStatus);
+                    return null;
+                default:
+                    return null;
+            }
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+            logger.warn("Exception when calling {}", url);
+            return null;
+        }
+    }
+
+    public static @Nullable List<SensorDataValue> getLatestValues(String response) {
+        Gson gson = new Gson();
+        SensorData[] valueArray = gson.fromJson(response, SensorData[].class);
+        if (valueArray.length == 0) {
+            return null;
+        } else if (valueArray.length == 1) {
+            SensorData v = valueArray[0];
+            return v.getSensordatavalues();
+        } else if (valueArray.length > 1) {
+            // declare first item as latest
+            SensorData latestData = valueArray[0];
+            String latestTimeStr = latestData.getTimeStamp();
+            Date latestTime = DateTimeUtils.toDate(latestTimeStr);
+            if (latestTime == null) {
+                logDateConversionError(response, latestData);
+            }
+            for (int i = 1; i < valueArray.length; i++) {
+                SensorData iterData = valueArray[i];
+                String iterTimeStr = iterData.getTimeStamp();
+                Date iterTime = DateTimeUtils.toDate(iterTimeStr);
+                if (iterTime == null) {
+                    logDateConversionError(response, latestData);
+                }
+                if (iterTime != null && latestTime != null) {
+                    if (latestTime.before(iterTime)) {
+                        // found item is newer - take it as latest
+                        latestTime = iterTime;
+                        latestData = iterData;
+                    } // else - found item is older - nothing to do
+
+                } else {
+                    logger.warn("One or two dates cannot be decoded 1) {} 2) {}", iterTimeStr, latestTimeStr);
+                }
+            }
+            return latestData.getSensordatavalues();
+        } else {
+            return null;
+        }
+    }
+
+    public static void logDateConversionError(final String response, final Object dto) {
+        logger.warn("Unable to get timestamp");
+        logger.warn("Response: {}", response);
+        Gson gson = new Gson();
+        String json = gson.toJson(dto);
+        logger.warn("GSon: {}", json);
+    }
+
+    public static boolean isParticulate(@Nullable List<SensorDataValue> valueList) {
+        if (valueList == null) {
+            return false;
+        }
+        Iterator<SensorDataValue> iter = valueList.iterator();
+        while (iter.hasNext()) {
+            SensorDataValue v = iter.next();
+            if (!(v.getValue_type().equals(P1) || v.getValue_type().equals(P2))) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
You could have some fun with streams here. You could try something like this. Same for below.
```suggestion
        return valueList.stream().map(v -> v.getValue_type()).filter(t -> t.equals(P1) || t.equals(P2)).findAny().isPresent();
```

Reply:
Nice - changed for all isXYZ functions